src/quorumnet/sn_network.cpp

```diff
@@ -0,0 +1,1019 @@
+    }
+}
+
+void SNNetwork::expire_idle_peers() {
+    for (auto &peer : peers) {
+        auto &info = peer.second;
+        auto idle = info.last_activity - std::chrono::steady_clock::now();
+        if (idle <= info.idle_expiry)
+            continue;
+        auto outgoing = info.outgoing.lock();
+        if (!outgoing)
+            continue;
+
+        SN_LOG(info, "Closing connection to " << as_hex(peer.first) << ": idle timeout reached");
+
+        for (size_t i = 0; i < remotes.size(); i++) {
+            if (remotes[i] == outgoing) {
+                remotes[i]->setsockopt<int>(ZMQ_LINGER, 0);
+                remotes.erase(remotes.begin() + i);
+                pollitems.erase(pollitems.begin() + poll_remote_offset + i);
+                assert(remotes.size() == pollitems.size() + poll_remote_offset);
+                break;
+            }
+        }
+    }
+}
+
+void SNNetwork::proxy_loop(const std::vector<std::string> &bind) {
+    zmq::socket_t zap_auth{context, zmq::socket_type::rep};
+    zap_auth.setsockopt<int>(ZMQ_LINGER, 0);
+    zap_auth.bind(ZMQ_ADDR_ZAP);
+
+    workers.setsockopt<int>(ZMQ_ROUTER_MANDATORY, 1);
+    workers.bind(SN_ADDR_WORKERS);
+
+    spawn_worker("w1");
+    int next_worker_id = 2;
+
+    // Set up the public tcp listener(s):
+    auto &listen = *listener;
+    listen.setsockopt(ZMQ_ZAP_DOMAIN, AUTH_DOMAIN_SN, sizeof(AUTH_DOMAIN_SN)-1);
+    listen.setsockopt<int>(ZMQ_CURVE_SERVER, 1);
+    listen.setsockopt(ZMQ_CURVE_PUBLICKEY, pubkey.data(), pubkey.size());
+    listen.setsockopt(ZMQ_CURVE_SECRETKEY, privkey.data(), privkey.size());
+//    listen.setsockopt<int>(ZMQ_ROUTER_HANDOVER, 1);
+    listen.setsockopt<int>(ZMQ_ROUTER_MANDATORY, 1);
+
+    for (const auto &b : bind) {
+        SN_LOG(info, "Quorumnet listening on " << b);
+        listen.bind(b);
+    }
+
+    // Also add an internal connection to self so that calling code can avoid needing to
+    // special-case rare situations where we are supposed to talk to a quorum member that happens to
+    // be ourselves (which can happen, for example, with cross-quoum Blink communication)
+    self_listener->bind(SN_ADDR_SELF);
+
+    add_pollitem(command);
+    add_pollitem(workers);
+    add_pollitem(zap_auth);
+    assert(pollitems.size() == poll_internal_size);
+    add_pollitem(*listener);
+    add_pollitem(*self_listener);
+    assert(pollitems.size() == poll_remote_offset);
+
+    constexpr auto poll_timeout = 5000ms; // Maximum time we spend in each poll
+    constexpr auto timeout_check_interval = 2000ms; // Minimum time before for checking for connections to close since the last check
+    auto last_conn_timeout = std::chrono::steady_clock::now();
+
+    std::string waiting_for_worker; // If set contains the identify of a worker we just spawned but haven't received a READY signal from yet
+    size_t last_conn_index = 0; // Index of the connection we last received a message from; see below
+
+    while (true) {
```

Comment:
> I'm thinking that this should use threadpool, so as to respect the max thread count the user specifies.

The max threads is really the number of worker threads that actually do anything intensive; the seperate messaging processing thread is always needed, and the vast majority of the time this thread is simply waiting on a poll.  There will never be anything CPU intensive or blocking in this thread, by design.

> In particular being able to queue the processing of an incoming message means you don't have this, I poll zmq- then find I have an incoming message, I need to spawn a thread and redo the loop awkwardness because you need to wait for it to be ready- where as you can queue the socket initialisation + handling of message in the one go.

This new-worker code only happens up to `max_workers - 1` times total in the lifetime of the SNNetwork; it's triggered if there is still unused room for new workers when a request arrives (i.e. we have only spawned 3 workers before but all are currently busy and we are allowed 4 or more).  The idea here being that we never spawn worker threads until there is actual demand for them, so even if you are running with max_threads = hardware_concurrency it's fairly unlikely you will actually spawn that many workers.

As to why avoiding doing message handling in worker threads (either directly or via a queue): that would require a mutex to read/write from it: a worker locks the mutex, polls for an incoming message (or checks the queue), then unlocks the mutex and processes the message.  Avoiding any mutexes in the communication layer is a design decision here: mutexes are *slow* and get slower the more you use them as contention rises (which is why lokid chokes with more than a few dozen p2p peers), while passing messages via local ZMQ inproc sockets has no such contention.  ZMQ is designed from the ground up to be lock-free and to enable lock-free multithreading communication.  There's a whole bunch more detail in here, but the above is the gist of it:

http://zguide.zeromq.org/page:all#Multithreading-with-ZeroMQ

Reply:
I see. Where is it that worker threads get closed at the moment other than on program exit? Also where do they get returned to be an idle worker? 

I don't think any of the monero code uses lockless paradigms, so yes agreed traditional mutex on heavily shared data structures suck, using atomics for lockless data structures are the only performant option. Though I see they advise heavily against it.

I'm just in favour of making the code path more linear if we can because linear flow of execution is easier to grasp which would be moving out that needing to wait for thread to get ready.

Reply:
> Where is it that worker threads get closed at the moment other than on program exit?

They don't, but if there are no jobs those threads are just going to be idling stuck on a socket poll that never arrives.

> Also where do they get returned to be an idle worker?

In `worker_thread` the `send_control(sock, "READY");` is the thread ready signal.